### PR TITLE
Add Optimization Algorithms

### PR DESCRIPTION
**Pull request by whatbeg**
## What changes were proposed in this pull request?

Added 4 optimization algorithms, Adam, Adadelta, RMSprop, Adamax.

## How was this patch tested?

Unit test passed，mainly applyed these optimization algorithms on Rosenbrock function，testing performance，and compared them with Torch's corresponding algorothms。
Code Style check passed.
ready to test performances.

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/516

### CODE REVIEW

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/Adadelta.scala [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    *                  state("accDelta"): vector of accumulated delta of gradients
+    * @return the new x vector and the function list {fx}, evaluated before the update
+    */
+  override def optimize(feval: (Tensor[T]) => (T, Tensor[T]),
+               parameter: Tensor[T], config: Table, state: Table): (Tensor[T], Array[T]) = {
+
+    val _config = if (config == null) T() else config
+    val _state = if (state == null) _config else state
+
+    val nevals = _state.get[Int]("evalCounter").getOrElse(0)
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/Adadelta.scala]: _state.getOrElse

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/Adadelta.scala [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      if (_state.get[Tensor[T]]("paramVariance").isDefined) {
+        (_state.get[Tensor[T]]("paramVariance").get, _state.get[Tensor[T]]("paramStd").get,
+          _state.get[Tensor[T]]("delta").get, _state.get[Tensor[T]]("accDelta").get)
+      } else {
+        (Tensor[T]().resizeAs(dfdx).zero(), Tensor[T]().resizeAs(dfdx).zero(),
+          Tensor[T]().resizeAs(dfdx).zero(), Tensor[T]().resizeAs(dfdx).zero())
+      }
+
+    _paramVariance.mul(ev.fromType[Double](dr)).addcmul(ev.fromType[Double](1-dr), dfdx, dfdx)
+    _paramStd.resizeAs(_paramVariance).copy(_paramVariance).add(ev.fromType[Double](eps)).sqrt()
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/Adadelta.scala]: why we need the resizeAs here? _paramVariance and _paramStd should be same size with dfdx, right?

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/Adam.scala [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    *                  state("denom"): A tmp tensor to hold the sqrt(v) + epsilon
+    * @return the new x vector and the function list {fx}, evaluated before the update
+    */
+  override def optimize(feval: (Tensor[T]) => (T, Tensor[T]),
+               parameter: Tensor[T], config: Table, state: Table): (Tensor[T], Array[T]) = {
+
+    val _config = if (config == null) T() else config
+    val _state = if (state == null) _config else state
+
+    val lr = _config.get[Double]("learningRate").getOrElse(1e-3)
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/Adam.scala]: _config.getOrElse

**Comment by yiheng** [comment]
Please run a code style check via 
mvn clean verify -DskipTests

**Comment by whatbeg** [comment]
@yiheng  OK，will do it soon. and check what you commented above :)

**Comment by yiheng** [comment]
please merge and revise the commit messages

**Comment by whatbeg** [comment]
@yiheng  have merged and revised commit messages : )

**Pull request merged**